File: core/web-ui/backend/app/api/modules.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import List, Optional, Dict, Any
from datetime import datetime

router = APIRouter()


class ModuleStatus(BaseModel):
    """モジュールステータス"""
    name: str
    version: str
    status: str  # running, stopped, error
    description: str
    metrics: Optional[Dict[str, Any]] = None
    last_update: datetime


class ModuleList(BaseModel):
    """モジュール一覧"""
    modules: List[ModuleStatus]


@router.get("", response_model=ModuleList)
async def get_modules():
    """
    導入済みモジュール一覧を取得

    Returns:
        ModuleList: モジュールのリスト
    """
    # TODO: Docker SDKを使って実際のコンテナ情報を取得
    # 現在はダミーデータを返す
    return ModuleList(
        modules=[
            ModuleStatus(
                name="farminpocket-temp",
                version="1.0.0",
                status="running",
                description="温度・湿度監視",
                metrics={
                    "temperature": 24.5,
                    "humidity": 65.0
                },
                last_update=datetime.now()
            ),
            ModuleStatus(
                name="farminpocket-irrigation",
                version="1.0.0",
                status="running",
                description="自動潅水制御",
                metrics={
                    "next_watering": "2025-11-17T14:30:00Z",
                    "last_watering": "2025-11-17T12:30:00Z"
                },
                last_update=datetime.now()
            )
        ]
    )


@router.get("/{module_name}", response_model=ModuleStatus)
async def get_module_detail(module_name: str):
    """
    モジュール詳細を取得

    Args:
        module_name: モジュール名

    Returns:
        ModuleStatus: モジュールの詳細情報
    """
    # TODO: 実際のモジュール情報を取得
    if module_name == "farminpocket-temp":
        return ModuleStatus(
            name="farminpocket-temp",
            version="1.0.0",
            status="running",
            description="温度・湿度監視",
            metrics={
                "temperature": 24.5,
                "humidity": 65.0
            },
            last_update=datetime.now()
        )
    else:
        raise HTTPException(status_code=404, detail="Module not found")


@router.post("/{module_name}/start")
async def start_module(module_name: str):
    """
    モジュールを起動

    Args:
        module_name: モジュール名

    Returns:
        Dict: 操作結果
    """
    # TODO: Docker SDKを使ってコンテナを起動
    return {"message": f"Module {module_name} started successfully"}


@router.post("/{module_name}/stop")
async def stop_module(module_name: str):
    """
    モジュールを停止

    Args:
        module_name: モジュール名

    Returns:
        Dict: 操作結果
    """
    # TODO: Docker SDKを使ってコンテナを停止
    return {"message": f"Module {module_name} stopped successfully"}


@router.post("/{module_name}/restart")
async def restart_module(module_name: str):
    """
    モジュールを再起動

    Args:
        module_name: モジュール名

    Returns:
        Dict: 操作結果
    """
    # TODO: Docker SDKを使ってコンテナを再起動
    return {"message": f"Module {module_name} restarted successfully"}

File: core/web-ui/frontend/src/stores/modules.js
import { defineStore } from 'pinia'
import axios from 'axios'

export const useModulesStore = defineStore('modules', {
  state: () => ({
    modules: [],
    loading: false,
    error: null
  }),

  actions: {
    async fetchModules() {
      this.loading = true
      this.error = null
      try {
        const response = await axios.get('/api/modules')
        this.modules = response.data.modules
      } catch (error) {
        this.error = error.message
        console.error('Failed to fetch modules:', error)
      } finally {
        this.loading = false
      }
    },

    async startModule(moduleName) {
      try {
        await axios.post(`/api/modules/${moduleName}/start`)
        await this.fetchModules()
      } catch (error) {
        console.error('Failed to start module:', error)
        throw error
      }
    },

    async stopModule(moduleName) {
      try {
        await axios.post(`/api/modules/${moduleName}/stop`)
        await this.fetchModules()
      } catch (error) {
        console.error('Failed to stop module:', error)
        throw error
      }
    },

    async restartModule(moduleName) {
      try {
        await axios.post(`/api/modules/${moduleName}/restart`)
        await this.fetchModules()
      } catch (error) {
        console.error('Failed to restart module:', error)
        throw error
      }
    }
  },

  getters: {
    runningModules: (state) => state.modules.filter(m => m.status === 'running'),
    stoppedModules: (state) => state.modules.filter(m => m.status === 'stopped'),
    errorModules: (state) => state.modules.filter(m => m.status === 'error')
  }
})
